### PR TITLE
IE Extensions table for Windows - Browser Helper Objects

### PR DESCRIPTION
**Pull request by ryanheffernan**
First attempt at getting IE extensions. This just gets Browser Helper Objects right now, and currently is just getting the reg/file paths. But this will be extended to include more types of extensions as we mature.

### CODE REVIEW

**Comment by osqueryer** [comment]
:no_good: The commit 57082c502921dd56c4a4d62d67ac421bdcd3f74a (Job results: [1913](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1913/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit 57082c502921dd56c4a4d62d67ac421bdcd3f74a (Job results: [5154](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5154/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 57082c502921dd56c4a4d62d67ac421bdcd3f74a (Job results: [2042](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2042/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 57082c502921dd56c4a4d62d67ac421bdcd3f74a (Job results: [2043](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2043/)) failed one or more tests (**Windows**).

File: osquery/filesystem/windows/fileops.cpp [hunk 1]
```diff
@@ -98,6 +98,43 @@ class WindowsFindFiles {
   fs::path path_;
 };
 
+Status windowsShortPathToLongPath(const std::string& shortPath,
+                                  std::string& rLongPath) {
+  TCHAR longPath[MAX_PATH];
```
Comment (theopolis) [inline osquery/filesystem/windows/fileops.cpp]: You could use `std::array<TCHAR>(MAX_PATH);` then `std::move` it into the string. Or do the old `&string[0]` trick to access a non-const buffer.

File: osquery/filesystem/windows/fileops.cpp [hunk 2]
```diff
@@ -98,6 +98,43 @@ class WindowsFindFiles {
...
+    return Status(1, "Failed to malloc for version info");
+  }
+  auto err = GetFileVersionInfo(path.c_str(), handle, verSize, verInfo.get());
+  if (err == 0) {
+    return Status(GetLastError(), "Failed to get file version info");
+  }
+  VS_FIXEDFILEINFO* pFileInfo = nullptr;
+  UINT verInfoSize = 0;
+  err = VerQueryValue(
+      verInfo.get(), TEXT("\\"), (LPVOID*)&pFileInfo, &verInfoSize);
```
Comment (theopolis) [inline osquery/filesystem/windows/fileops.cpp]: Can you just pass a `nullptr` for the 4th argument?

File: osquery/tables/system/windows/ie_extensions.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    for (const auto& exec : executables) {
+      Row r;
+      std::string fullPath;
+      std::string version;
+      ret = windowsShortPathToLongPath(exec, fullPath);
+      if (ret.ok()) {
+        r["path"] = fullPath;
+      }
+      ret = windowsGetFileVersion(exec, version);
```
Comment (theopolis) [inline osquery/tables/system/windows/ie_extensions.cpp]: nit, initialize your `version` here to make the definition/use tighter.

File: osquery/tables/system/windows/ie_extensions.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      LOG(WARNING) << "Failed to get class executables: " + ret.getMessage();
+      return ret;
+    }
+    for (const auto& exec : executables) {
+      Row r;
+      std::string fullPath;
+      std::string version;
+      ret = windowsShortPathToLongPath(exec, fullPath);
+      if (ret.ok()) {
+        r["path"] = fullPath;
```
Comment (theopolis) [inline osquery/tables/system/windows/ie_extensions.cpp]: You can `std::move` the contents into the `Row`.

File: osquery/tables/system/windows/ie_extensions.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      }
+      ret = windowsGetFileVersion(exec, version);
+      if (ret.ok()) {
+        r["version"] = version;
```
Comment (theopolis) [inline osquery/tables/system/windows/ie_extensions.cpp]: Same move recommendation.

File: osquery/filesystem/windows/fileops.cpp (hunk 1 above)
Comment (ryanheffernan) [inline osquery/filesystem/windows/fileops.cpp]: I'm not sure I follow? A std::string can't take ownership of a buffer (afaict), and I need a writable char buffer for GetLongPathName to populate. If I use a std::array I'm still going to have to pass a raw pointer to it to the Windows API, and I'm still going to have to construct a new std::string to return.

File: osquery/filesystem/windows/fileops.cpp (hunk 2 above)
Comment (ryanheffernan) [inline osquery/filesystem/windows/fileops.cpp]: The spec does not indicate this is an optional param (https://msdn.microsoft.com/en-us/library/windows/desktop/ms647464(v=vs.85).aspx), although when I tested it with a null pointer it still worked. But I'd prefer to not make the assumption :).

**Comment by osqueryer** [comment]
:-1: The commit bbe5d76838ba2e59b69f0ca32adea66c324df775 (Job results: [1304](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1304/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:no_good: The commit bbe5d76838ba2e59b69f0ca32adea66c324df775 (Job results: [1942](https://jenkins.osquery.io/job/osqueryPullRequestAudit/1942/)) failed the code audit or documentation test.

**Comment by osqueryer** [comment]
:-1: The commit bbe5d76838ba2e59b69f0ca32adea66c324df775 (Job results: [1305](https://jenkins.osquery.io/job/osqueryPullRequestOSX/1305/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit bbe5d76838ba2e59b69f0ca32adea66c324df775 (Job results: [5186](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5186/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit bbe5d76838ba2e59b69f0ca32adea66c324df775 (Job results: [5187](https://jenkins.osquery.io/job/osqueryPullRequestLinux/5187/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit bbe5d76838ba2e59b69f0ca32adea66c324df775 (Job results: [198](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/198/)) failed one or more tests (**FreeBSD**).

**Comment by osqueryer** [comment]
:-1: The commit bbe5d76838ba2e59b69f0ca32adea66c324df775 (Job results: [2084](https://jenkins.osquery.io/job/osqueryPullRequestWindows/2084/)) failed one or more tests (**Windows**).

**Pull request merged**